### PR TITLE
Fix json_expr simplification of pointers

### PR DESCRIPTION
**Pull request by smowton**
Improve json_expr's ability to figure out where pointers lead 

Previously it could understand "&symbol.member" but not "&symbol", a shortcoming that has been exposed by the solver now preferring the latter when the expression's type indicates that is more appropriate than a pointer to a base class or member. It also couldn't look through typecasts, which prevented it from following array data pointers (which would use the construct "(data = &((void*)data_array[0])").

Also add a test category for JSON traces, and an initial test checking for this patch's improvements.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -96,6 +96,7 @@ jobs:
             - g++-5
             - libubsan0
             - parallel
+            - jq
```
Comment (tautschnig) [inline .travis.yml:99]: Could we also have this in the buildspec* files, please?

File: src/goto-programs/json_expr.cpp [hunk 2]
```diff
@@ -46,27 +46,39 @@ static exprt simplify_json_expr(const exprt &src)
...
-    to_constant_expr(to_index_expr(to_address_of_expr(src).object()).index())
-      .value_is_zero_string())
+  else if(src.id() == ID_address_of)
   {
-    // simplify expressions of the form  &array[0]
-    return simplify_json_expr(
-      to_index_expr(to_address_of_expr(src).object()).array());
+    exprt object = to_address_of_expr(src).object();
+    while(object.id() == ID_typecast)
+      object = to_typecast_expr(object).op();
```
Comment (tautschnig) [inline src/goto-programs/json_expr.cpp]: How about `const exprt &object = skip_typecast(to_address_of_expr(src).object());`

File: src/goto-programs/json_expr.cpp [hunk 3]
```diff
@@ -46,27 +46,39 @@ static exprt simplify_json_expr(const exprt &src)
...
+    if(
+      object.id() == ID_member &&
+      id2string(
+        to_member_expr(to_address_of_expr(src).object()).get_component_name())
```
Comment (tautschnig) [inline src/goto-programs/json_expr.cpp]: `to_member_expr(object)`

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 4481da9f1555792a6398a236a950ebeeb8b648b8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97686747
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by smowton** [comment]
@tautschnig all done as requested

**Comment by smowton** [comment]
Looks like Joelbot's sadness is to do with a current test-gen/cbmc desync (therefore unrelated to this PR)

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d1419530a935c5bf2b6b402aa994394b6a21babc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97724442
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**